juju/juju
# state: update storage constraints and config in SetCharm

Application.SetCharm now supports updating
storage constraints and charm config settings
at the same time as updating an application's
charm.

Because you can now specify constraints for
new stores added to a charm, we relax the
rules around changes to a charm's stores. If
a new required store is added, we will add
max(min-count, constraint-count) instances
to each unit, where min-count is the minimum
count specified in the charm metadata, and
constraint-count is the user-specified count,
if any. If constraint-count < min-count, we
will return an error explaining that the
constraints do not satisfy the requirements.

The minimum count for existing stores may
also be changed now. If any unit does not
already have enough storage instances to
satisfy this requirement, new instances will
be added so long as the storage constraints
allow it.

If config is specified, we take the existing
settings, and delete any keys that are no
longer defined by the charm's config; we
then add the user-specified settings.

**QA**

(With https://github.com/juju/juju/pull/6312)

1. bootstrap and deploy a charm with a
   single optional store, with count=1
2. modify the charm to make said store
   required, with a minimum count of 2;
   and add another store with a minimum
   count of 3
3. upgrade-charm to said charm, without
   specifying --storage. Observe that it
   fails because the existing store's
   constraints (count=1) do not satisfy
   the new minimum (count>=2).
4. upgrade-charm to said charm, with
   --storage existing-store=2, but without
   specifying anything for the new store.
   This will succeed; there will be an
   additional storage instance for the
   existing store, and three instances
   for the new store with the default
   storage constraints.
5. Do-over, but this time with --storage for the new store also, specifying new-storage=4. Observe that the user-specified constraints are used for the new store, and four instances are created.
6. deploy a charm with one string config setting "foo".
7. remove "foo" from the charm, and add "bar". Upgrade to it without specifying any --config on the command line. We should have the existing behaviour: "foo" is removed, and the default value for "bar" is added.
8. add "foo" back to the charm, and remove "bar". Upgrade to it, but this time specify --config on the command line, pointing at a file that contains a non-default value for "foo". The application should have the value for "foo" taken from the config file, and "bar" removed.
9. change the value for "foo" in the config file, add a value for "bar", and run the upgrade again. This should fail, as there is no "bar" config defined by the charm. Application config should not have changed.
